pytorch_lightning/plugins/training_type/fully_sharded.py

```diff
@@ -0,0 +1,206 @@
+                unless using mixed precision, in which case defaults to torch.float16.
+                (Default: None).
+            bucket_cap_mb: bucket parameters so that gradient reduction
+                can potentially overlap with backward computation.
+                bucket_cap_mb controls the bucket size in MegaBytes (MB).
+                Buckets are sub-divided based on world_size,
+                so the max shard size is roughly bucket_cap_mb / world_size.
+                Values <= 0 disable bucketing.
+                (Default: 25).
+            min_num_params: Number of parameters to wrap when using FairScale ``auto_wrap``.
+                (Default: 1e8)
+            state_dict_to_cpu: Whether to return parameters (returned by :func:`state_dict`) on CPU device.
+                If ``False``, this will default to ``compute_device``.
+                (Defautl: True).
+        """
+
+        super().__init__(
+            parallel_devices=parallel_devices,
+            cluster_environment=cluster_environment,
+        )
+        self.cpu_offload = cpu_offload
+        self.move_grads_to_cpu = move_grads_to_cpu
+        self.flatten_parameters = flatten_parameters
+        self.reshard_after_forward = reshard_after_forward
+        self.fp32_reduce_scatter = fp32_reduce_scatter
+        self.compute_dtype = compute_dtype
+        self.bucket_cap_mb = bucket_cap_mb
+        self.min_num_params = min_num_params
+        self.state_dict_device = torch.device("cpu") if state_dict_to_cpu else None
+        self._process_group = None
+
+    @property
+    def process_group(self):
+        if self._process_group is None:
+            self._process_group = torch.distributed.new_group()
+        return self._process_group
+
+    def setup_distributed(self):
+        super().setup_distributed()
+        if self.root_device.type == "cuda":
+            torch.cuda.set_device(self.root_device)
+
+    @contextlib.contextmanager
+    def model_sharded_context(self) -> Generator:
+        precision = self.lightning_module.trainer.precision
+
+        def wrap_policy(*args, **kwargs):
+            return default_auto_wrap_policy(*args, **kwargs, min_num_params=self.min_num_params)
+
+        with enable_wrap(
+            wrapper_cls=FullyShardedDataParallel,
+            auto_wrap_policy=wrap_policy,
+            process_group=self.process_group,
+            cpu_offload=self.cpu_offload,
+            move_grads_to_cpu=self.move_grads_to_cpu,
+            flatten_parameters=self.flatten_parameters,
+            mixed_precision=precision == "mixed",
+            reshard_after_forward=self.reshard_after_forward,
+            fp32_reduce_scatter=self.fp32_reduce_scatter,
+            compute_dtype=self.compute_dtype,
+            bucket_cap_mb=self.bucket_cap_mb,
+            state_dict_device=self.state_dict_device,
+        ):
+            yield
+
+    def connect(self, model: Module) -> None:
```

Comment:
this is needed when we load from ckpt, as now we checkpoint the full state, to load, we need the unwrapped model first, and load. This allows us to configure shard (i.e. wrap) afterwards.

Added unittest for making sure that for different stages start (on_x_start),  `_assert_layer_fsdp_instance` passes.